### PR TITLE
gh-112087: Make list_{concat, repeat, inplace_repeat, ass_item)  to be thread-safe

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112087

### CODE REVIEW

**Comment by corona10** [comment]
Two PRs will be left.
1. list_ass_item related changes: Quite huge; I will submit the change separately.
2. QSBR-related changes. list_clear and list_new should use this technique.

File: Objects/listobject.c [hunk 1]
```diff
@@ -563,20 +556,12 @@ PyList_GetSlice(PyObject *a, Py_ssize_t ilow, Py_ssize_t ihigh)
 }
 
 static PyObject *
-list_concat(PyObject *aa, PyObject *bb)
+list_concat_locked(PyListObject *a, PyListObject *b)
```
Comment (colesbury) [inline Objects/listobject.c]: Let's use a consistent naming convention for the functions that assume the caller holds the lock. I think @mpage and @DinoV have been using "_holds_lock".

**Review by colesbury — APPROVED** [review]
Looks good, but I'd prefer we use consistent naming schemes for the functions that assume the callers hold the lock.

File: Objects/listobject.c (hunk 1 above)
Comment (corona10) [reply Objects/listobject.c]: Okay, since other implementations are using the `xxxx_lock_held` convention, let's follow it.

**Pull request merged**